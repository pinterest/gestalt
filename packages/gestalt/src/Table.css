:root {
  /* primary colors */
  --gestalt-colorGray0: #fff;
}

.table {
  border-collapse: separate;
  border-spacing: 0;
  width: 100%;
}

.th {
  composes: alignLeft from "./Typography.css";
  composes: paddingY3 from "./boxWhitespace.css";
  composes: paddingX3 from "./boxWhitespace.css";
}

.td {
  composes: paddingY3 from "./boxWhitespace.css";
  composes: paddingX3 from "./boxWhitespace.css";
}

.sticky tr th {
  background-color: var(--gestalt-colorGray0);
  position: sticky;
  top: 0;
}

.thead tr:last-child th {
  border-bottom: 1px solid #ddd;
}
