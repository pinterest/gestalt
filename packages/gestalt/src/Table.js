// @flow strict
import * as React from 'react';
import Box from './Box.js';
import styles from './Table.css';
import TableCell from './TableCell.js';
import TableBody from './TableBody.js';
import TableFooter from './TableFooter.js';
import TableHeader from './TableHeader.js';
import TableHeaderCell from './TableHeaderCell.js';
import TableRow from './TableRow.js';
import TableSortableHeaderCell from './TableSortableHeaderCell.js';

type Props = {|
  children: React.Node,
  borderSize?: 'sm' | 'none',
|};

export default function Table(props: Props) {
  const { borderSize, children } = props;

  return (
    <Box
      overflow="scrollX"
      {...(borderSize === 'sm' ? { borderSize: 'sm', rounding: 1 } : {})}
    >
      <table className={styles.table}>{children}</table>
    </Box>
  );
}

Table.Body = TableBody;

Table.Cell = TableCell;

Table.Footer = TableFooter;

Table.Header = TableHeader;

Table.HeaderCell = TableHeaderCell;

Table.Row = TableRow;

Table.SortableHeaderCell = TableSortableHeaderCell;
