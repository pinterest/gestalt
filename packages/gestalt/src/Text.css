/*
This file was generated by running

  $ /Users/chrislloyd/.nvm/versions/node/v9.2.1/bin/node /Users/chrislloyd/src/gestalt/scripts/text.js

*/

:root {
  --font-size-1: 12px;
  --font-size-2: 14px;
  --font-size-3: 16px;
  --font-size-4: 18px;
  --font-size-5: 21px;
}

.Text {
  composes: antialiased sansSerif letterSpacing from "./Typography.css";
}

.fontSize1 {
  font-size: var(--font-size-1);
}

.fontSize2 {
  font-size: var(--font-size-2);
}

.fontSize3 {
  font-size: var(--font-size-3);
}

.fontSize4 {
  font-size: var(--font-size-4);
}

.fontSize5 {
  font-size: var(--font-size-5);
}

@media (--sm) {
  .smFontSize1 {
    font-size: var(--font-size-1);
  }

  .smFontSize2 {
    font-size: var(--font-size-2);
  }

  .smFontSize3 {
    font-size: var(--font-size-3);
  }

  .smFontSize4 {
    font-size: var(--font-size-4);
  }

  .smFontSize5 {
    font-size: var(--font-size-5);
  }
}

@media (--md) {
  .mdFontSize1 {
    font-size: var(--font-size-1);
  }

  .mdFontSize2 {
    font-size: var(--font-size-2);
  }

  .mdFontSize3 {
    font-size: var(--font-size-3);
  }

  .mdFontSize4 {
    font-size: var(--font-size-4);
  }

  .mdFontSize5 {
    font-size: var(--font-size-5);
  }
}

@media (--lg) {
  .lgFontSize1 {
    font-size: var(--font-size-1);
  }

  .lgFontSize2 {
    font-size: var(--font-size-2);
  }

  .lgFontSize3 {
    font-size: var(--font-size-3);
  }

  .lgFontSize4 {
    font-size: var(--font-size-4);
  }

  .lgFontSize5 {
    font-size: var(--font-size-5);
  }
}
