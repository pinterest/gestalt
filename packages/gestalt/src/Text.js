// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import colors from './Colors.css';
import styles from './Text.css';
import typography from './Typography.css';

const SIZE_SCALE: { [size: ?string]: number } = {
  xs: 1,
  sm: 2,
  md: 3,
  lg: 4,
  xl: 5,
};

type Props = {|
  align?: 'left' | 'right' | 'center' | 'justify',
  bold?: boolean,
  children?: React.Node,
  color?:
    | 'green'
    | 'pine'
    | 'olive'
    | 'blue'
    | 'navy'
    | 'midnight'
    | 'purple'
    | 'orchid'
    | 'eggplant'
    | 'maroon'
    | 'watermelon'
    | 'orange'
    | 'darkGray'
    | 'gray'
    | 'lightGray'
    | 'red'
    | 'white',
  inline?: boolean,
  italic?: boolean,
  overflow?: 'normal' | 'breakWord',
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  smSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  mdSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  lgSize?: 'xs' | 'sm' | 'md' | 'lg' | 'xl',
  leading?: 'tall' | 'short',
  truncate?: boolean,
  __dangerouslyIncreaseLineHeight?: boolean,
|};

export default function Text({
  align = 'left',
  bold = false,
  children,
  color = 'darkGray',
  inline = false,
  italic = false,
  overflow = 'breakWord',
  size = 'md',
  smSize,
  mdSize,
  lgSize,
  leading,
  truncate = false,
  __dangerouslyIncreaseLineHeight = false,
}: Props) {
  const scale = SIZE_SCALE[size];
  const smScale = SIZE_SCALE[smSize];
  const mdScale = SIZE_SCALE[mdSize];
  const lgScale = SIZE_SCALE[lgSize];

  const cs = cx(
    styles.Text,
    styles[`fontSize${scale}`],
    lgSize && styles[`lgFontSize${lgScale}`],
    mdSize && styles[`mdFontSize${mdScale}`],
    smSize && styles[`smFontSize${smScale}`],
    color === 'blue' && colors.blue,
    color === 'darkGray' && colors.darkGray,
    color === 'eggplant' && colors.eggplant,
    color === 'gray' && colors.gray,
    color === 'green' && colors.green,
    color === 'lightGray' && colors.lightGray,
    color === 'maroon' && colors.maroon,
    color === 'midnight' && colors.midnight,
    color === 'navy' && colors.navy,
    color === 'olive' && colors.olive,
    color === 'orange' && colors.orange,
    color === 'orchid' && colors.orchid,
    color === 'pine' && colors.pine,
    color === 'purple' && colors.purple,
    color === 'red' && colors.red,
    color === 'watermelon' && colors.watermelon,
    color === 'white' && colors.white,
    leading === 'short' && typography.leadingShort,
    (leading === 'tall' || __dangerouslyIncreaseLineHeight) &&
      typography.leadingTall,
    align === 'center' && typography.alignCenter,
    align === 'justify' && typography.alignJustify,
    align === 'left' && typography.alignLeft,
    align === 'right' && typography.alignRight,
    overflow === 'breakWord' && typography.breakWord,
    italic && typography.fontStyleItalic,
    !italic && typography.fontStyleNormal,
    bold && typography.fontWeightBold,
    !bold && typography.fontWeightNormal,
    truncate && typography.truncate
  );
  const Tag = inline ? 'span' : 'div';

  return (
    <Tag
      className={cs}
      {...(truncate && typeof children === 'string'
        ? { title: children }
        : null)}
    >
      {children}
    </Tag>
  );
}

Text.propTypes = {
  __dangerouslyIncreaseLineHeight: PropTypes.bool,
  align: PropTypes.oneOf(['left', 'right', 'center', 'justify']),
  bold: PropTypes.bool,
  children: PropTypes.node,
  color: PropTypes.oneOf([
    'green',
    'pine',
    'olive',
    'blue',
    'navy',
    'midnight',
    'purple',
    'orchid',
    'eggplant',
    'maroon',
    'watermelon',
    'orange',
    'darkGray',
    'gray',
    'lightGray',
    'red',
    'white',
  ]),
  inline: PropTypes.bool,
  italic: PropTypes.bool,
  leading: PropTypes.oneOf(['tall', 'short']),
  lgSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  mdSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  overflow: PropTypes.oneOf(['normal', 'breakWord']),
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  smSize: PropTypes.oneOf(['xs', 'sm', 'md', 'lg', 'xl']),
  truncate: PropTypes.bool,
};
