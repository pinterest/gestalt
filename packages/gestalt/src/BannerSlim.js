// @flow strict
import { Children, type Element, Fragment, type Node as ReactNode } from 'react';
import Box from './Box';
import Button from './Button';
import ButtonLink from './ButtonLink';
import { useDefaultLabelContext } from './contexts/DefaultLabelProvider';
import Flex from './Flex';
import Icon from './Icon';
import IconButton from './IconButton';
import Link from './Link';
import MESSAGING_TYPE_ATTRIBUTES from './MESSAGING_TYPE_ATTRIBUTES';
import Text from './Text';

type DismissButtonType = {
  accessibilityLabel?: string,
  onDismiss: () => void,
};

function DismissButton({ accessibilityLabel, onDismiss }: DismissButtonType) {
  const { accessibilityDismissButtonLabel } = useDefaultLabelContext('BannerSlim');
  return (
    <IconButton
      accessibilityLabel={accessibilityLabel ?? accessibilityDismissButtonLabel}
      icon="cancel"
      iconColor="darkGray"
      onClick={onDismiss}
      size="xs"
    />
  );
}

type HelperLinkType = {
  accessibilityLabel: string,
  href: string,
  onClick?: ({
    event: SyntheticMouseEvent<HTMLAnchorElement> | SyntheticKeyboardEvent<HTMLAnchorElement>,
    dangerouslyDisableOnNavigation: () => void,
  }) => void,
  target?: null | 'self' | 'blank',
  text: string,
};

function HelperLink({ accessibilityLabel, href, onClick, target, text }: HelperLinkType) {
  return (
    <Text inline>
      <Link
        accessibilityLabel={accessibilityLabel}
        display="inlineBlock"
        href={href}
        onClick={onClick}
        target={target}
      >
        {text}
      </Link>
    </Text>
  );
}

type PrimaryActionType =
  | {
      accessibilityLabel: string,
      disabled?: boolean,
      href: string,
      label: string,
      onClick?: $ElementType<React$ElementConfig<typeof ButtonLink>, 'onClick'>,
      rel?: 'none' | 'nofollow',
      role: 'link',
      target?: null | 'self' | 'blank',
    }
  | {
      accessibilityLabel: string,
      disabled?: boolean,
      label: string,
      onClick: $ElementType<React$ElementConfig<typeof Button>, 'onClick'>,
      role?: 'button',
    };

function PrimaryAction({ accessibilityLabel, disabled, label, ...props }: PrimaryActionType) {
  return props.role === 'link' ? (
    <ButtonLink
      accessibilityLabel={accessibilityLabel}
      color="white"
      disabled={disabled}
      fullWidth
      href={props.href}
      onClick={props.onClick}
      rel={props.rel}
      size="sm"
      target={props.target}
      text={label}
    />
  ) : (
    <Button
      accessibilityLabel={accessibilityLabel}
      color="white"
      disabled={disabled}
      fullWidth
      onClick={props.onClick}
      size="sm"
      text={label}
    />
  );
}

type Props = {
  /**
   * Adds a dismiss button to BannerSlim. See the [Dismissible variant](https://gestalt.pinterest.systems/web/bannerslim#Dismissible) for more info.
   * The `accessibilityLabel` should follow the [Accessibility guidelines](https://gestalt.pinterest.systems/web/bannerslim#Accessibility).
   *
   * Note that compact ("___Bare" type) BannerSlims are not dismissable.
   */
  dismissButton?: DismissButtonType,
  /**
   * Helper [Link](https://gestalt.pinterest.systems/web/link) to be placed after the message. See the [Message variant](https://gestalt.pinterest.systems/web/bannerslim#Message) to learn more.
   */
  helperLink?: HelperLinkType,
  /**
   * Label to describe the status iconâ€™s purpose. See the [Accessibility guidelines](https://gestalt.pinterest.systems/web/bannerslim#Accessibility) for details on proper usage.
   */
  iconAccessibilityLabel?: string,
  /**
   * Main content of BannerSlim. Content should be [localized](https://gestalt.pinterest.systems/web/bannerslim#Localization). See the [Message variant](https://gestalt.pinterest.systems/web/bannerslim#Message) to learn more.
   *
   */
  message: string | Element<typeof Text>,
  /**
   * Main action for users to take on BannerSlim. If `href` is supplied, the action will serve as a link. See [GlobalEventsHandlerProvider](https://gestalt.pinterest.systems/web/utilities/globaleventshandlerprovider#Link-handlers) to learn more about link navigation.
   * If no `href` is supplied, the action will be a button.
   * The `accessibilityLabel` should follow the [Accessibility guidelines](https://gestalt.pinterest.systems/web/bannerslim#Accessibility).
   * See the [Primary action](https://gestalt.pinterest.systems/web/bannerslim#Primary-action) variant to learn more.
   *
   * Note that actions are not available on compact ("___Bare" type) BannerSlims.
   */
  primaryAction?: PrimaryActionType,
  /**
   * The type of BannerSlim. See the [variants](https://gestalt.pinterest.systems/web/bannerslim#Variants) to learn more.
   */
  type?:
    | 'neutral'
    | 'error'
    | 'info'
    | 'warning'
    | 'success'
    | 'recommendation'
    | 'errorBare'
    | 'infoBare'
    | 'warningBare'
    | 'successBare'
    | 'recommendationBare',
};

/**
 * [BannerSlim](https://gestalt.pinterest.systems/web/bannerslim) conveys brief information related to a specific section of a page. The message can relay success, warning, error or general information. Since they are about a specific section of a page or surface, BannerSlim sits inside of a container, and not at the top of the page. For alerts that apply to the whole page, use [BannerCallout](https://gestalt.pinterest.systems/web/bannercallout).
 *
 * ![BannerSlim light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/BannerSlim.spec.mjs-snapshots/BannerSlim-chromium-darwin.png)
 * ![BannerSlim dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/BannerSlim-dark.spec.mjs-snapshots/BannerSlim-dark-chromium-darwin.png)
 *
 */
export default function BannerSlim({
  dismissButton,
  helperLink,
  iconAccessibilityLabel,
  message,
  primaryAction,
  type = 'neutral',
}: Props): ReactNode {
  const isBare = type.endsWith('Bare');
  const isDefault = type === 'neutral';
  const { backgroundColor, iconColor, icon } = MESSAGING_TYPE_ATTRIBUTES[type.replace('Bare', '')];
  const {
    iconAccessibilityLabelError,
    iconAccessibilityLabelInfo,
    iconAccessibilityLabelRecommendation,
    iconAccessibilityLabelSuccess,
    iconAccessibilityLabelWarning,
  } = useDefaultLabelContext('BannerSlim');

  const getDefaultIconAccessibilityLabel = () => {
    switch (type) {
      case 'success':
      case 'successBare':
        return iconAccessibilityLabelSuccess;
      case 'info':
      case 'infoBare':
        return iconAccessibilityLabelInfo;
      case 'recommendation':
      case 'recommendationBare':
        return iconAccessibilityLabelRecommendation;
      case 'warning':
      case 'warningBare':
        return iconAccessibilityLabelWarning;
      case 'error':
      case 'errorBare':
        return iconAccessibilityLabelError;
      default:
        return '';
    }
  };
  // Buttons not allowed on compact BannerSlims
  const shouldShowButtons = !isBare && (primaryAction || dismissButton);

  return (
    <Box
      alignItems="center"
      color={isBare ? 'transparent' : backgroundColor}
      direction="column"
      display="flex"
      mdDirection="row"
      padding={isBare ? 0 : 4}
      paddingY={isBare ? 1 : 0}
      rounding={4}
      width="100%"
    >
      <Flex
        alignItems="start"
        flex="grow"
        gap={{ row: isBare ? 2 : 4, column: 0 }}
        mdAlignItems="center"
        width="100%"
      >
        {!isDefault && (
          <Flex.Item alignSelf={shouldShowButtons ? undefined : 'start'}>
            <Icon
              accessibilityLabel={iconAccessibilityLabel ?? getDefaultIconAccessibilityLabel()}
              color={iconColor}
              icon={icon}
              size={16}
            />
          </Flex.Item>
        )}

        <Flex.Item flex="grow">
          <Box
            dangerouslySetInlineStyle={{
              __style: !isDefault ? { marginTop: '-1px' } : {},
            }}
          >
            {typeof message === 'string' ? (
              <Text inline>
                {message}
                {helperLink ? (
                  <Fragment>
                    {' '}
                    <HelperLink {...helperLink} />
                  </Fragment>
                ) : null}
              </Text>
            ) : null}
            {typeof message !== 'string' &&
            Children.only<Element<typeof Text>>(message).type.displayName === 'Text'
              ? message
              : null}
          </Box>
        </Flex.Item>

        {shouldShowButtons && (
          <Flex.Item flex="none">
            <Flex alignItems="center" gap={{ row: 4, column: 0 }}>
              {primaryAction && (
                <Box display="none" flex="none" mdDisplay="flex">
                  <PrimaryAction {...primaryAction} />
                </Box>
              )}

              {dismissButton && <DismissButton {...dismissButton} />}
            </Flex>
          </Flex.Item>
        )}
      </Flex>
      {!isBare && primaryAction && (
        <Box alignSelf="end" display="flex" flex="none" marginTop={4} mdDisplay="none">
          <PrimaryAction {...primaryAction} />
        </Box>
      )}
    </Box>
  );
}

BannerSlim.displayName = 'BannerSlim';
