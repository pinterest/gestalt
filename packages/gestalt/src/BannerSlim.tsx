import { Children, ComponentProps, Fragment, ReactElement } from 'react';
import Box from './Box';
import Button from './Button';
import ButtonLink from './ButtonLink';
import { useDefaultLabelContext } from './contexts/DefaultLabelProvider';
import Flex from './Flex';
import Icon from './Icon';
import IconButton from './IconButton';
import Link from './Link';
import MESSAGING_TYPE_ATTRIBUTES from './MESSAGING_TYPE_ATTRIBUTES';
import Text from './Text';

type HelperLinkType = {
  accessibilityLabel: string;
  href: string;
  onClick?: (arg1: {
    event: React.MouseEvent<HTMLAnchorElement> | React.KeyboardEvent<HTMLAnchorElement>;
    dangerouslyDisableOnNavigation: () => void;
  }) => void;
  target?: null | 'self' | 'blank';
  text: string;
};

function HelperLink({ accessibilityLabel, href, onClick, target, text }: HelperLinkType) {
  return (
    <Text inline>
      <Link
        accessibilityLabel={accessibilityLabel}
        display="inlineBlock"
        href={href}
        onClick={onClick}
        target={target}
      >
        {text}
      </Link>
    </Text>
  );
}

type PrimaryActionType =
  | {
      accessibilityLabel: string;
      disabled?: boolean;
      href: string;
      label: string;
      onClick?: ComponentProps<typeof ButtonLink>['onClick'];
      rel?: 'none' | 'nofollow';
      role: 'link';
      target?: null | 'self' | 'blank';
    }
  | {
      accessibilityLabel: string;
      disabled?: boolean;
      label: string;
      onClick: ComponentProps<typeof Button>['onClick'];
      role?: 'button';
    };

function PrimaryAction({ accessibilityLabel, disabled, label, ...props }: PrimaryActionType) {
  return props.role === 'link' ? (
    <ButtonLink
      accessibilityLabel={accessibilityLabel}
      color="white"
      disabled={disabled}
      fullWidth
      href={props.href}
      onClick={props.onClick}
      rel={props.rel}
      size="sm"
      target={props.target}
      text={label}
    />
  ) : (
    <Button
      accessibilityLabel={accessibilityLabel}
      color="white"
      disabled={disabled}
      fullWidth
      onClick={props.onClick}
      size="sm"
      text={label}
    />
  );
}

type Props = {
  /**
   * Helper [Link](https://gestalt.pinterest.systems/web/link) to be placed after the message. See the [Message variant](https://gestalt.pinterest.systems/web/bannerslim#Message) to learn more.
   */
  helperLink?: HelperLinkType;
  /**
   * Label to describe the status iconâ€™s purpose. See the [Accessibility guidelines](https://gestalt.pinterest.systems/web/bannerslim#Accessibility) for details on proper usage.
   */
  iconAccessibilityLabel?: string;
  /**
   * Main content of BannerSlim. Content should be [localized](https://gestalt.pinterest.systems/web/bannerslim#Localization). See the [Message variant](https://gestalt.pinterest.systems/web/bannerslim#Message) to learn more.
   *
   */
  message: string | ReactElement;
  /**
   * Adds a dismiss button to BannerSlim. Uses default accessibility label. See the [Dismiss button](https://gestalt.pinterest.systems/web/bannerslim#Dismiss-button) variant to learn more.
   */
  onDismiss?: () => void;
  /**
   * Main action for users to take on BannerSlim. If `href` is supplied, the action will serve as a link. See [GlobalEventsHandlerProvider](https://gestalt.pinterest.systems/web/utilities/globaleventshandlerprovider#Link-handlers) to learn more about link navigation.
   * If no `href` is supplied, the action will be a button.
   * The `accessibilityLabel` should follow the [Accessibility guidelines](https://gestalt.pinterest.systems/web/bannerslim#Accessibility).
   * See the [Primary action](https://gestalt.pinterest.systems/web/bannerslim#Primary-action) variant to learn more.
   *
   * Note that actions are not available on compact ("___Bare" type) BannerSlims.
   */
  primaryAction?: PrimaryActionType;
  /**
   * The type of BannerSlim. See the [variants](https://gestalt.pinterest.systems/web/bannerslim#Variants) to learn more.
   */
  type?:
    | 'neutral'
    | 'error'
    | 'info'
    | 'warning'
    | 'success'
    | 'recommendation'
    | 'errorBare'
    | 'infoBare'
    | 'warningBare'
    | 'successBare'
    | 'recommendationBare';
};

/**
 * [BannerSlim](https://gestalt.pinterest.systems/web/bannerslim) conveys brief information related to a specific section of a page. The message can relay success, warning, error or general information. Since they are about a specific section of a page or surface, BannerSlim sits inside of a container, and not at the top of the page. For alerts that apply to the whole page, use [BannerCallout](https://gestalt.pinterest.systems/web/bannercallout).
 *
 * ![BannerSlim light mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/BannerSlim.spec.ts-snapshots/BannerSlim-chromium-darwin.png)
 * ![BannerSlim dark mode](https://raw.githubusercontent.com/pinterest/gestalt/master/playwright/visual-test/BannerSlim-dark.spec.ts-snapshots/BannerSlim-dark-chromium-darwin.png)
 *
 */
export default function BannerSlim({
  helperLink,
  iconAccessibilityLabel,
  message,
  onDismiss,
  primaryAction,
  type = 'neutral',
}: Props) {
  const isBare = type.endsWith('Bare');
  const isDefault = type === 'neutral';
  // @ts-expect-error - TS7053 - Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'Readonly<{ neutral: { backgroundColor: string; }; success: { icon: string; iconColor: string; color: string; backgroundColor: string; }; info: { icon: string; iconColor: string; color: string; backgroundColor: string; }; warning: { ...; }; error: { ...; }; recommendation: { ...; }; }>'.
  const { backgroundColor, iconColor, icon } = MESSAGING_TYPE_ATTRIBUTES[type.replace('Bare', '')];
  const { accessibilityDismissButtonLabel } = useDefaultLabelContext('BannerSlim');
  const {
    iconAccessibilityLabelError,
    iconAccessibilityLabelInfo,
    iconAccessibilityLabelRecommendation,
    iconAccessibilityLabelSuccess,
    iconAccessibilityLabelWarning,
  } = useDefaultLabelContext('BannerSlim');

  const getDefaultIconAccessibilityLabel = () => {
    switch (type) {
      case 'success':
      case 'successBare':
        return iconAccessibilityLabelSuccess;
      case 'info':
      case 'infoBare':
        return iconAccessibilityLabelInfo;
      case 'recommendation':
      case 'recommendationBare':
        return iconAccessibilityLabelRecommendation;
      case 'warning':
      case 'warningBare':
        return iconAccessibilityLabelWarning;
      case 'error':
      case 'errorBare':
        return iconAccessibilityLabelError;
      default:
        return '';
    }
  };

  // Buttons not allowed on compact BannerSlims
  const shouldShowButtons = !isBare && (primaryAction || onDismiss);

  return (
    <Box
      alignItems="center"
      color={isBare ? 'transparent' : backgroundColor}
      direction="column"
      display="flex"
      mdDirection="row"
      padding={isBare ? 0 : 4}
      paddingY={isBare ? 1 : 0}
      rounding={4}
      width="100%"
    >
      <Flex
        alignItems="start"
        flex="grow"
        gap={{ row: isBare ? 2 : 4, column: 0 }}
        mdAlignItems="center"
        width="100%"
      >
        {!isDefault && (
          <Flex.Item alignSelf={shouldShowButtons ? undefined : 'start'}>
            <Icon
              accessibilityLabel={iconAccessibilityLabel ?? getDefaultIconAccessibilityLabel()}
              color={iconColor}
              icon={icon}
              size={16}
            />
          </Flex.Item>
        )}

        <Flex.Item flex="grow">
          <Box
            dangerouslySetInlineStyle={{
              __style: !isDefault ? { marginTop: '-1px' } : {},
            }}
          >
            {typeof message === 'string' ? (
              <Text inline>
                {message}
                {helperLink ? (
                  <Fragment>
                    {' '}
                    <HelperLink {...helperLink} />
                  </Fragment>
                ) : null}
              </Text>
            ) : null}
            {typeof message !== 'string' &&
            // @ts-expect-error - TS2339
            Children.only<ReactElement>(message).type.displayName === 'Text'
              ? message
              : null}
          </Box>
        </Flex.Item>

        {shouldShowButtons && (
          <Flex.Item flex="none">
            <Flex alignItems="center" gap={{ row: 4, column: 0 }}>
              {primaryAction && (
                <Box display="none" flex="none" mdDisplay="flex">
                  <PrimaryAction {...primaryAction} />
                </Box>
              )}

              {onDismiss && (
                <IconButton
                  accessibilityLabel={accessibilityDismissButtonLabel}
                  icon="cancel"
                  iconColor="darkGray"
                  onClick={onDismiss}
                  size="xs"
                />
              )}
            </Flex>
          </Flex.Item>
        )}
      </Flex>
      {!isBare && primaryAction && (
        <Box alignSelf="end" display="flex" flex="none" marginTop={4} mdDisplay="none">
          <PrimaryAction {...primaryAction} />
        </Box>
      )}
    </Box>
  );
}

BannerSlim.displayName = 'BannerSlim';
