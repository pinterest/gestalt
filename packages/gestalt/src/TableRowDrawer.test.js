// @flow strict
import renderer from 'react-test-renderer';
import BannerSlim from './BannerSlim';
import Table from './Table';
import Text from './Text';

test('renders correctly with drawer', () => {
  const tree = renderer
    .create(
      <Table accessibilityLabel="Table Row Drawer">
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>
              <Text>TestA</Text>
            </Table.HeaderCell>
            <Table.HeaderCell>
              <Text>TestB</Text>
            </Table.HeaderCell>
            <Table.HeaderCell>
              <Text>TestC</Text>
            </Table.HeaderCell>
            <Table.HeaderCell>
              <Text>TestD</Text>
            </Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>
          <Table.RowDrawer
            drawerContents={
              <BannerSlim
                iconAccessibilityLabel="test"
                message="Test BannerSlim"
                type="recommendation"
              />
            }
            id="drawerTest"
          >
            <Table.Cell>
              <Text>Row1A</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row1B</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row1C</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row1D</Text>
            </Table.Cell>
          </Table.RowDrawer>
          <Table.Row>
            <Table.Cell>
              <Text>Row2A</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row2B</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row2C</Text>
            </Table.Cell>
            <Table.Cell>
              <Text>Row2D</Text>
            </Table.Cell>
          </Table.Row>
        </Table.Body>
      </Table>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
