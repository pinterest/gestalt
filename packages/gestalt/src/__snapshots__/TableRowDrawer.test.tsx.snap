// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with drawer 1`] = `
<div
  className="box overflowAuto"
  style={
    Object {
      "maxHeight": undefined,
    }
  }
  tabIndex={0}
>
  <table
    className="table"
  >
    <caption
      className="absolute box overflowHidden"
      style={
        Object {
          "clip": "rect(1px, 1px, 1px, 1px)",
          "height": 1,
          "whiteSpace": "nowrap",
          "width": 1,
        }
      }
    >
      Table Row Drawer
    </caption>
    <thead
      className="thead"
    >
      <tr
        className=""
      >
        <th
          className="th"
          scope="col"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            TestA
          </div>
        </th>
        <th
          className="th"
          scope="col"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            TestB
          </div>
        </th>
        <th
          className="th"
          scope="col"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            TestC
          </div>
        </th>
        <th
          className="th"
          scope="col"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            TestD
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      className="tbody"
    >
      <tr
        aria-details="drawerTest"
        className=""
      >
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row1A
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row1B
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row1C
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row1D
          </div>
        </td>
      </tr>
      <tr
        id="drawerTest"
      >
        <td
          className="drawer"
          colSpan={4}
        >
          <div
            className="box paddingX2 paddingY2"
          >
            <div
              className="box mdDirectionRow paddingX4 paddingY0 paddingY4 recommendationWeak rounding4 xsDirectionColumn xsDisplayFlex xsItemsCenter"
              style={
                Object {
                  "width": "100%",
                }
              }
            >
              <div
                className="Flex rowGap4 columnGap0 flexGrow mdItemsCenter xsDirectionRow xsItemsStart"
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <div
                  className="FlexItem selfStart"
                >
                  <svg
                    aria-hidden={null}
                    aria-label="test"
                    className="recommendation icon iconBlock"
                    height={16}
                    role="img"
                    viewBox="0 0 24 24"
                    width={16}
                  >
                    <path
                      d="test-file-stub"
                    />
                  </svg>
                </div>
                <div
                  className="FlexItem flexGrow"
                >
                  <div
                    className="box"
                    style={
                      Object {
                        "marginTop": "-1px",
                      }
                    }
                  >
                    <span
                      className="default alignStart breakWord Text fontSize300 fontWeightNormal"
                    >
                      Test BannerSlim
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
      <tr
        className=""
      >
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row2A
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row2B
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row2C
          </div>
        </td>
        <td
          className="td"
          style={
            Object {
              "left": undefined,
              "right": undefined,
            }
          }
        >
          <div
            className="default alignStart breakWord Text fontSize300 fontWeightNormal"
          >
            Row2D
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
